src/suffix.c

```diff
@@ -124,55 +128,211 @@ void deleteSuffixTrie(Trie *trie, const char *str, uint32_t len) {
+      continue;
+    }
+
+    // 1. long string are likely to have less results
+    // 2. tokens at end of pattern are likely to be more relevant
+    int curScore = tokenLen[i] + i;
+
+    // iterating all children is demanding
+    if (str[tokenIdx[i] + tokenLen[i]] == '*') {
+      curScore -= 5;
+    }
+
+    // this branching is heavy
+    for (int j = tokenIdx[i]; j < tokenIdx[i] + tokenLen[i]; ++j) {
+      if (str[j] == '?') {
+        --curScore;
+      }
+    }
+
+    if (curScore >= score) {
+      score = curScore;
+      retidx = i;
+    }
+  }
+
+  return retidx;
+}
+
+int Suffix_ChooseToken_rune(const rune *str, size_t len, size_t *tokenIdx, size_t *tokenLen) {
+  int runner = 0;
+  int i = 0;
+  int init = 0;
+  while (i < len) {
+    if (str[i] != (rune)'*') {
+      tokenIdx[runner] = i;
+      init = 1;
+    }
+    while (i < len && str[i] != (rune)'*') {
+      ++i;
+    }
+    if (init) {
+      tokenLen[runner] = i - tokenIdx[runner];
+      ++runner;
+    }
+    while (str[i] == (rune)'*') {
+      ++i;
+    }
+  }
+
+  // choose best option
+  int score = INT32_MIN;
+  int retidx = -1;
+  for (int i = 0; i < runner; ++i) {
+    if (tokenLen[i] < MIN_SUFFIX) {
+      continue;
+    }
+
+    // 1. long string are likely to have less results
+    // 2. tokens at end of pattern are likely to be more relevant
+    int curScore = tokenLen[i] + 1;
+
+    // iterating all children is demanding
+    if (str[tokenIdx[i] + tokenLen[i]] == (rune)'*') {
+      curScore -= 5;
+    }
+
+    // this branching is heavy
+    for (int j = tokenIdx[i]; j < tokenIdx[i] + tokenLen[i]; ++j) {
+      if (str[j] == (rune)'?') {
+        --score;
+      }
     }
-  }                              
+
+    if (curScore >= score) {
+      score = curScore;
+      retidx = i;
+    }
+  }
+
+  return retidx;
+}
+
+static int processSuffixData_Wildcard(suffixData *data, SuffixCtx *sufCtx) {
+  if (!data) {
+    return REDISMODULE_OK;
+  }
+  arrayof(char *) array = data->array;
+  for (int i = 0; i < array_len(array); ++i) {
+    if (Wildcard_MatchChar(sufCtx->cstr, sufCtx->cstrlen, array[i], strlen(array[i]))
+            == FULL_MATCH) {
+      if (sufCtx->callback(array[i], strlen(array[i]), sufCtx->cbCtx, NULL) != REDISMODULE_OK) {
+        return REDISEARCH_ERR;
+      }
+    }
+  }
+  return REDISMODULE_OK;
+}
+
+int Suffix_CB_Wildcard(const rune *rune, size_t len, void *p, void *payload) {
+  SuffixCtx *sufCtx = p;
+  TriePayload *pl = payload;
+
+  return processSuffixData_Wildcard((void *)pl->data, sufCtx);
+}
+
+void Suffix_IterateWildcard(SuffixCtx *sufCtx) {
+  size_t idx[sufCtx->cstrlen];
+  size_t lens[sufCtx->cstrlen];
+  int useIdx = Suffix_ChooseToken_rune(sufCtx->rune, sufCtx->runelen, idx, lens);
+
+  if (useIdx == -1) {
+    return;
+  }
+
+  rune *token = sufCtx->rune + idx[useIdx];
+  size_t toklen = lens[useIdx];
+  if (token[toklen] == (rune)'*') {
+    toklen++;
+  }
+  token[toklen] = (rune)'\0';
```

Comment:
Are we the owner of this string that it is OK to change it here? In other wards, do we have a copy of the srting that we keep on `sufCtx`?

Reply:
Yes. We copy it in `NewWildcardNode_WithParams`.